examples/rbm_image_generation.py

```diff
@@ -0,0 +1,214 @@
+# Copyright 2025 D-Wave
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import torch
+from torch.utils.data import DataLoader
+from dwave.plugins.torch.models.boltzmann_machine import (
+    RestrictedBoltzmannMachine as RBM,
+)
+from torchvision import transforms, datasets
+import matplotlib.pyplot as plt
+
+
+def load_binarized_mnist(dataset_path: str = "data") -> datasets.MNIST:
+    """
+    Load the MNIST dataset and binarize it (pixels >= 0.5 become 1, else 0).
+
+    Args:
+        dataset_path (str): Path to download/store the MNIST dataset. Defaults to "data".
+
+    Returns:
+        datasets.MNIST: Binarized MNIST training dataset.
+    """
+    transform = transforms.Compose(
+        [transforms.ToTensor(), transforms.Lambda(lambda x: (x >= 0.5).float())]
+    )
+
+    train_dataset = datasets.MNIST(
+        root=dataset_path, train=True, transform=transform, download=True
+    )
+    return train_dataset
+
+
+def train_loop(
+    train_loader: DataLoader,
+    rbm: RBM,
+    n_epochs: int,
+    n_gibbs_steps: int,
+    learning_rate: float,
+    momentum: float,
+    weight_decay: float,
+) -> None:
+    """
+    Train the RBM using contrastive divergence with momentum and weight decay.
+
+    Args:
+        train_loader (DataLoader): PyTorch DataLoader for training data.
+        rbm (RBM): Restricted Boltzmann Machine instance.
+        n_epochs (int): Number of training epochs.
+        n_gibbs_steps (int): Number of Gibbs sampling steps per CD update.
+        learning_rate (float): Base learning rate.
+        momentum (float): Momentum coefficient for parameter updates.
+        weight_decay (float): Weight decay (L2 regularization) coefficient.
+    """
+    device = rbm._weights.device
+    for epoch in range(n_epochs):
+        total_error = 0
+        num_examples = 0
+        for batch, _ in train_loader:
+            # flatten input data
+            batch = batch.reshape(batch.size(0), rbm.n_visible).to(device)
+
+            # Perform one step of contrastive divergence and accumulate error
+            error = rbm._contrastive_divergence(
```

Comment:
If `_contrastive_divergence` is to be used outside of the class (as it is here), it shouldn't be hidden.